Fading out non compatible datasources prototype 3

lorem--ipsum opened the pull request:
Made a PR to spin up an ephemeral env.

lorem--ipsum committed:
third proto

lorem--ipsum committed:
no tabs

lorem--ipsum committed:
refresh icon is now hidden unless hovered

lorem--ipsum committed:
reorganized and cleaned up the data importer code

lorem--ipsum committed:
a bit more renaming

lorem--ipsum committed:
fixup

lorem--ipsum committed:
final reorg -> more stuff in data importer

lorem--ipsum committed:
closes VIZ-604 tooltip for long names

lorem--ipsum committed:
closes VIZ-602 cannot remove bubble size

lorem--ipsum committed:
fixup (smarter removal of columns)

lorem--ipsum committed:
closes VIZ-603 additional compatibility for pies and scalars

lorem--ipsum committed:
closes VIZ-601 no waterfall in visualizer after all

eliotmetabase commented on frontend/src/metabase/visualizer/components/DataImporter/DatasetsList/DatasetsListItem.tsx:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    if (currentDisplay === "scalar") {
+      return display === "scalar";
```
Instead of checking the viz type, what do you think about this logic:
- If the starting data set only has 1 field, then it's compatible with data sets that (a) have 1 field and (b) has the same column type (ie number -> number)
- If the starting data has more than 1 field (ie. a timeseries), then it's never compatible with data sets that only have 1 field (ie. a scalar)

lorem--ipsum replied:
Yeah that sounds reasonable

eliotmetabase commented on frontend/src/metabase/visualizer/components/DataImporter/DatasetsList/DatasetsListItem.tsx:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    if (!primaryColumn || !metadata) {
+      return true;
```
Is this checking if the starting data set has no primary column or any fields selected, and we're defaulting to compatible for any other data sets? 
If so - I would argue this should be `false`. Otherwise, what are we mapping against?

lorem--ipsum replied:
No that's for when no dataset is currently selected

eliotmetabase commented on frontend/src/metabase/visualizer/components/DataImporter/DatasetsList/DatasetsListItem.tsx:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    if (isNumber(primaryColumn) || isString(primaryColumn)) {
+      return fields?.some(field => field.id === primaryColumn.id);
```
From some testing, this could lead to an interesting case where the ID from 2 different fields in 2 different tables match, even though they're not the same.
Should we match on name instead of id? Or both? 
Just id caused some weird bugs in my testing.

lorem--ipsum replied:
A combination of both would probably be more restrictive, which is what we want IMO.

eliotmetabase commented on frontend/src/metabase/visualizer/components/DataImporter/DatasetsList/DatasetsListItem.tsx:
```diff
@@ -0,0 +1,133 @@
+    getFieldsMetadata(item.id);
+  }, [item, getFieldsMetadata]);
+
+  const isCompatible = useMemo(() => {
```
Scatter plots seem to act weird, since BE adds an extra field to the data set. I think we should avoid those in visualizations for now.

lorem--ipsum replied:
So like, make theme _never_ compatible?

kulyk replied:
We could also filter out these columns when checking for compatibility. Not sure about scatter plots, but for pivot tables we could do something like that:

```ts
import { isPivotGroupColumn } from "metabase/lib/data_grid"

cols.filter(col => !isPivotGroupColumn(col))`
```

eliotmetabase reviewed (commented):
@lorem--ipsum Conor and I were playing around with compatibility, and refined some of the rules. I left comments, let me know what you think.

eliotmetabase commented on frontend/src/metabase/visualizer/components/DataImporter/DatasetsList/DatasetsListItem.tsx:
```diff
@@ -0,0 +1,133 @@
+      return fields?.some(field => field.id === primaryColumn.id);
+    }
+
+    return fields?.some(field => field.base_type === primaryColumn?.base_type);
```
For dates, should we check if the granularity matches, or is at least similar enough?
As per a comment from Kyle's figjam: 
> Time field with ability to resolve time range.
> Month + Year works
> Day of Week and Month, wouldn’t work

lorem--ipsum replied:
Yes

eliotmetabase commented:
Added a workflow to describe the rules of compatibility as I understand them: https://www.figma.com/board/Ao4eCPJgetbwmKqvfBgxJW/Visualizer-next-steps?node-id=28-157&t=lGzBroOvu7UgaSji-4

Let me know what you think!
cc @cdeweyx @lorem--ipsum @kdoh

lorem--ipsum committed:
visual improvements

lorem--ipsum committed:
implemented better matching algorithm

lorem--ipsum commented:
> Added a workflow to describe the rules of compatibility as I understand them: https://www.figma.com/board/Ao4eCPJgetbwmKqvfBgxJW/Visualizer-next-steps?node-id=28-157&t=lGzBroOvu7UgaSji-4
> 
> Let me know what you think! cc @cdeweyx @lorem--ipsum @kdoh

I pushed again and addressed your comments :-)

cdeweyx commented:
Minor but noticed that the "Info" popover for description isn't working yet, overall great stuff here

eliotmetabase reviewed (approved):
Stamping based on my comments being resolved

Please also get a review from another eng on the team :)

lorem--ipsum commented:
> Minor but noticed that the "Info" popover for description isn't working yet, overall great stuff here

Yes I wan't sure what it was supposed to do ^^

kulyk commented on frontend/src/metabase/visualizer/components/DataImporter/DatasetsList/DatasetsListItem.tsx:
```diff
@@ -0,0 +1,134 @@
+  const currentDisplay = useSelector(getVisualizationType);
+  const primaryColumn = useSelector(getVisualizerPrimaryColumn);
+
+  const [metadata, setMetadata] = useState<{
+    display?: VisualizationDisplay | undefined;
+    fields?: Field[] | undefined;
+  }>({});
+
+  const dispatch = useDispatch();
+
+  const getFieldsMetadata = useCallback(
+    async (id: VisualizerDataSourceId) => {
+      const { type, sourceId } = parseDataSourceId(id);
+      if (type === "card") {
+        const result = await dispatch(
+          cardApi.endpoints.getCard.initiate({ id: sourceId }),
+        );
+
+        setMetadata({
+          display: result.data?.display,
+          fields: result.data?.result_metadata,
+        });
+      }
+    },
+    [dispatch],
+  );
+
+  useEffect(() => {
+    getFieldsMetadata(item.id);
+  }, [item, getFieldsMetadata]);
```
WDYT about using `useGetCardQuery` from `metabase/api` here instead?

kulyk commented on frontend/src/metabase/visualizer/components/DataImporter/DatasetsList/getIsCompatible.ts:
```diff
@@ -0,0 +1,86 @@
+    return false;
+  }
+
+  if (fields.length <= 1) {
+    return false;
+  }
```
I believe it should return `true` if we're building a funnel and `targetDataset` is scalar's card dataset, as it'd have only one column

kulyk commented on frontend/src/metabase/visualizer/components/DataImporter/DatasetsList/getIsCompatible.ts:
IIRC, we shouldn't be looking for compatible columns among all `fields`, but only a subset filtered with `isMetric` and `isDimension` from `"metabase-lib/v1/types/utils/isa"` (because, for example, not all numeric columns can actually be used as a metric)

lorem--ipsum commented on frontend/src/metabase/visualizer/visualizer.slice.ts:
```diff
@@ -287,22 +288,26 @@ const visualizerHistoryItemSlice = createSlice({
     },
-    removeColumn: (state, action: PayloadAction<{ name: string }>) => {
-      const { name } = action.payload;
+    removeColumn: (
+      state,
+      action: PayloadAction<{ name: string; well?: "bubble" }>,
+    ) => {
```
Because maybe we're trying to remove a column from a different well but it also happens to be the one we use in the bubble

lorem--ipsum commented on frontend/src/metabase/visualizer/components/DataImporter/DatasetsList/getIsCompatible.ts:
```diff
@@ -0,0 +1,86 @@
+  const typeMatcher = (field: Field) =>
+    compareTypeAndName(primaryColumn, field);
+
+  if (currentDisplay === "scalar") {
+    return (
+      targetDataset?.display === "scalar" && !!fields?.find(idAndNameMatcher)
+    );
+  }
```
You're right, yeah this litterally takes the specs from Eliot's doc and translates them.

lorem--ipsum committed:
addressed comments

kulyk commented on frontend/src/metabase/visualizer/components/DataImporter/DatasetsList/getIsCompatible.ts:
```diff
@@ -0,0 +1,86 @@
+    return fields.some(idAndNameMatcher) || fields.some(typeMatcher);
+  }
+
+  if (isDate(primaryColumn)) {
+    return fields.some(field => {
+      return typeMatcher(field) && (field as any).unit === primaryColumn.unit;
+    });
+  }
```
We can plot series with different time units, so I think we can remove the `unit` check

lorem--ipsum commented on frontend/src/metabase/visualizer/components/DataImporter/DatasetsList/getIsCompatible.ts:
```diff
@@ -0,0 +1,86 @@
+  const typeMatcher = (field: Field) =>
+    compareTypeAndName(primaryColumn, field);
+
+  if (currentDisplay === "scalar") {
```
No, this addresses the case where we want to combine two scalars as described here: https://www.figma.com/board/Ao4eCPJgetbwmKqvfBgxJW/Visualizer-next-steps?node-id=28-157&t=kX94g85s7YLAoOod-0

lorem--ipsum committed:
addressed comments
